services/api/lib/postgre.js

```diff
@@ -295,6 +295,295 @@ module.exports = function (pg) {
+                  data.id
+                ];
+                break;
+              }
+              case 'remove': {
+                values = [
+                  data.id
+                ];
+                break;
+              }
+            }
+            break;
+          }
+          case 'pages': {
+            switch (method) {
+              case 'create': {
+                values = [
+                  data.projectId,
+                  userId,
+                  data.x,
+                  data.y,
+                  JSON.stringify(data.styles || {})
+                ];
+                break;
+              }
+              case 'update': {
+                values = [
+                  data.x,
+                  data.y,
+                  JSON.stringify(data.styles || {}),
+                  data.id
+                ];
+                break;
+              }
+              case 'remove': {
+                values = [
+                  data.id
+                ];
+                break;
+              }
+            }
+            break;
+          }
+          case 'elements': {
+            switch (method) {
+              case 'create': {
+                values = [
+                  data.pageId,
+                  userId,
+                  data.type,
+                  JSON.stringify(data.attributes || {}),
+                  JSON.stringify(data.styles || {})
+                ];
+                break;
+              }
+              case 'update': {
+                values = [
+                  JSON.stringify(data.styles || {}),
+                  JSON.stringify(data.attributes || {}),
+                  data.id
+                ];
+                break;
+              }
+              case 'remove': {
+                values = [
+                  data.id
+                ];
+                break;
+              }
+            }
+          }
+        }
+        return values;
+      }
+
+      function hasPermissionsForAction(userId, type, method, data, results, transaction) {
+        return new Promise(function(resolve, reject) {
+          var lookupType;
+          var lookupId;
+          if ( method === 'create' ) {
+            switch (type) {
+              case 'projects': {
+                return resolve(true);
+              }
+              case 'pages': {
+                lookupType = 'projects';
+                lookupId = data.projectId;
+                break;
+              }
+              case 'elements': {
+                lookupType = 'pages';
+                lookupId = data.pageId;
+                break;
+              }
+            }
+          } else {
+            lookupType = type;
+            lookupId = data.id;
+          }
+
+          // coerce these values to string, because bigints come in from PG as strings
+          lookupId = '' + lookupId;
+          userId = '' + userId;
+
+          var txResult;
+          for (var i = 0; i < results.length; ++i) {
```

Comment:
filter?
```
var txResult = results.filter(function(result) { return (...&&...&&...); })[0]
```
that way if there's a match, txResult will be that match. If not, it'll stay `undefined`